clients/da-vinci-client/src/main/java/com/linkedin/davinci/consumer/VeniceChangelogConsumerImpl.java

```diff
@@ -155,28 +150,140 @@ public VeniceChangelogConsumerImpl(
+            dictionary.array());
+      }
+    } else {
+      compressor = compressorFactory.getCompressor(version.getCompressionStrategy());
+    }
+    return compressor;
+  }
+
+  @Override
+  public CompletableFuture<Void> seekToBeginningOfPush(Set<Integer> partitions) {
+    // Get latest version topic
+    storeRepository.refresh();
+    Store store = storeRepository.getStore(storeName);
+    int currentVersion = store.getCurrentVersion();
+    String topic = Version.composeKafkaTopic(storeName, currentVersion);
+    return internalSeek(partitions, topic, kafkaConsumer::seekToBeginning);
+  }
+
+  @Override
+  public CompletableFuture<Void> seekToBeginningOfPush() {
+    return seekToBeginningOfPush(
+        kafkaConsumer.assignment()
+            .stream()
+            .map(topicPartition -> topicPartition.partition())
+            .collect(Collectors.toSet()));
+  }
+
+  @Override
+  public CompletableFuture<Void> seekToEndOfPush(Set<Integer> partitions) {
+    // Get latest change capture topic
+    storeRepository.refresh();
+    Store store = storeRepository.getStore(storeName);
+    int currentVersion = store.getCurrentVersion();
+    String topic = Version.composeKafkaTopic(storeName, currentVersion) + ChangeCaptureView.CHANGE_CAPTURE_TOPIC_SUFFIX;
+    return internalSeek(partitions, topic, kafkaConsumer::seekToBeginning);
+  }
+
+  @Override
+  public CompletableFuture<Void> seekToEndOfPush() {
+    return seekToEndOfPush(
+        kafkaConsumer.assignment()
+            .stream()
+            .map(topicPartition -> topicPartition.partition())
+            .collect(Collectors.toSet()));
+  }
+
+  @Override
+  public CompletableFuture<Void> seekToTail(Set<Integer> partitions) {
+    // Get latest change capture topic
+    String topic = getCurrentServingVersionTopic() + ChangeCaptureView.CHANGE_CAPTURE_TOPIC_SUFFIX;
+    return internalSeek(partitions, topic, kafkaConsumer::seekToEnd);
+  }
+
+  private String getCurrentServingVersionTopic() {
+    storeRepository.refresh();
+    Store store = storeRepository.getStore(storeName);
+    int currentVersion = store.getCurrentVersion();
+    return Version.composeKafkaTopic(storeName, currentVersion);
+  }
+
+  @Override
+  public CompletableFuture<Void> seekToTail() {
+    return seekToTail(
+        kafkaConsumer.assignment()
+            .stream()
+            .map(topicPartition -> topicPartition.partition())
+            .collect(Collectors.toSet()));
+  }
+
+  @Override
+  public CompletableFuture<Void> seekToCheckpoint(Set<VeniceChangeCoordinate> checkpoints) {
+    return CompletableFuture.supplyAsync(() -> {
+      for (VeniceChangeCoordinate coordinate: checkpoints) {
+        internalSeek(Collections.singleton(coordinate.getPartition()), coordinate.getTopic(), foo -> {
+          // TODO: This is a hack until we refactor out kafkaConsumer, delete this.
+          Long topicOffset = ((ApacheKafkaOffsetPosition) coordinate.getPosition()).getOffset();
```

Comment:
As for venice wrapper of kafka consumer: `ApacheKafkaConsumerAdapter`, we actually seek to +1 offset for read next one, do we need to follow that there or we are OK for reading exactlly from that offset?

Reply:
Good question.  That would make sense, however, there isn't really a way in the future to do that kind of math.  We can't do arithmetic on offsets on Pubsub positions.  So what I think will have to happen is that that kind of behavior will end up getting dictated to us by whatever implementation is going on under the hood.

So with that in mind, I'd rather not implement that kind of behavior here.

Reply:
Honestly I think it is ok so long as it's properly communicated.  IMO, getting served the write you asked for makes sense semantically, and has more utility realistically (for example, say you persist the offset of an event you are interested in reading again?  Being able to seek and read at that position is the only you could do it).

Reply:
I guess to give further testimony to doing this, from the samza api doc on the requirement for behvaior:  `offset - String representing the offset of the point in the stream to start reading messages from. This is an inclusive parameter; if "7" were specified, the first message for the system/stream/partition to be consumed and returned would be a message whose offset is "7".
`